### PR TITLE
Refactor: split up feature model anomaly solver into different anomaly types

### PR DESCRIPTION
**Pull request by schweikart**

### CODE REVIEW

**Review by Elscrux — CHANGES_REQUESTED** [review]
The separate feature model solvers are much better now. Since we are using the Webflux framework now, the only real overhead of this approach is the additional meta solver, which is valid and perhaps even desired.

**Comment by schweikart** [comment]
> the only real overhead of this approach is the additional meta solver, which is valid and perhaps even desired.

We might even be able to get rid of that. We could implement an ` abstract SingletonProblemSolver extends ProblemSolver implements MetaSolver` that implements the `MetaSolver` methods automatically. This could be used for problems with only a single solver.

File: src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/voidmodel/VoidFeatureMetaSolver.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import edu.kit.provideq.toolbox.meta.ProblemSolver;
+import edu.kit.provideq.toolbox.meta.ProblemType;
+import edu.kit.provideq.toolbox.meta.setting.MetaSolverSetting;
+import java.util.List;
+import org.springframework.stereotype.Component;
+
+@Component
+public class VoidFeatureMetaSolver
+    extends MetaSolver<String, String, ProblemSolver<String, String>> {
+  private final VoidFeatureSolver solver;
```
[the remark this answers is not in the document]
Comment (schweikart) [reply src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/voidmodel/VoidFeatureMetaSolver.java:14]: agreed, I struggled to come up with a precise name. `SatBasedVoidFeatureModelMetaSolver` would be precise but I'm not sure if it's too long...

File: src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/dead/DeadFeatureMetaSolver.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import edu.kit.provideq.toolbox.meta.MetaSolver;
+import edu.kit.provideq.toolbox.meta.Problem;
+import edu.kit.provideq.toolbox.meta.ProblemType;
+import edu.kit.provideq.toolbox.meta.setting.MetaSolverSetting;
+import java.util.List;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DeadFeatureMetaSolver extends MetaSolver<String, String, DeadFeatureSolver> {
+  private final DeadFeatureSolver solver = new DeadFeatureSolver();
```
[the remark this answers is not in the document]
Comment (schweikart) [reply src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/dead/DeadFeatureMetaSolver.java:12]: That wasn't intended, it's probably a result of IntelliJ's autocompletion... I'll change it

File: src/main/java/edu/kit/provideq/toolbox/meta/ProblemType.java [hunk 3]
```diff
@@ -29,7 +29,8 @@ public enum ProblemType {
    * <a href="https://sdq.kastel.kit.edu/publications/pdfs/kowal2016b.pdf">More information</a>
    */
   FEATURE_MODEL_ANOMALY("feature-model-anomaly", SolveFeatureModelRequest.class),
```
[the remark this answers is not in the document]
Comment (schweikart) [reply src/main/java/edu/kit/provideq/toolbox/meta/ProblemType.java:31]: nope, I'll remove it

File: src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/dead/DeadFeatureMetaSolver.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (schweikart) [reply src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/dead/DeadFeatureMetaSolver.java:12]: Maybe `SatBasedDeadFeatureMetaSolver`?

File: src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/voidmodel/VoidFeatureSolver.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    implements ProblemSolver<String, String> {
+  @Override
+  public String getName() {
+    return "SAT-based Void Feature Model Solver";
+  }
+
+  @Override
+  public List<SubRoutineDefinition> getSubRoutines() {
+    return List.of(
+        new SubRoutineDefinition(ProblemType.SAT, "sat",
```
[the remark this answers is not in the document]
Comment (schweikart) [reply src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/voidmodel/VoidFeatureSolver.java:28]: good point!

**Comment by Elscrux** [comment]
> We might even be able to get rid of that. We could implement an ` abstract SingletonProblemSolver extends ProblemSolver implements MetaSolver` that implements the `MetaSolver` methods automatically. This could be used for problems with only a single solver.

I'd rather keep it as it is, as a `SingletonProblemSolver` kind of goes against the point of a `MetaSolver` which should handle multiple solvers. We might not want to create another solver for something like a dead features anomaly in feature models, but at least we support it.

File: src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/dead/DeadFeatureMetaSolver.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Elscrux) [reply src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/dead/DeadFeatureMetaSolver.java:12]: Yeah, that's not very pretty but is more descriptive

File: src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/voidmodel/VoidFeatureMetaSolver.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Elscrux) [reply src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/voidmodel/VoidFeatureMetaSolver.java:14]: It's quite long, but I can't imagine anything else that's better. You could abbreviate some parts of the name (like FeatureModel -> FM), but that's also not very nice. So I'd say use that.

**Comment by schweikart** [comment]
I've fixed toolbox-web to use this new route scheme in https://github.com/ProvideQ/toolbox-web/pull/25

File: src/main/java/edu/kit/provideq/toolbox/meta/ProblemType.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Elscrux) [reply src/main/java/edu/kit/provideq/toolbox/meta/ProblemType.java:31]: I don't think this is fixed yet?

File: src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/voidmodel/VoidFeatureSolver.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Elscrux) [reply src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/voidmodel/VoidFeatureSolver.java:28]: I don't think this is fixed yet?

File: src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/dead/DeadFeatureMetaSolver.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Elscrux) [reply src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/dead/DeadFeatureMetaSolver.java:12]: I don't think this is fixed yet?

[the remark this answers is not in the document]
Comment (Elscrux) [reply src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/dead/DeadFeatureMetaSolver.java:12]: I don't think this is fixed yet?

[the remark this answers is not in the document]
Comment (schweikart) [reply src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/dead/DeadFeatureMetaSolver.java:12]: Now it is, I haven't gotten to revising this PR until now

**Comment by schweikart** [comment]
I've finally gotten to revising this PR now :sweat_smile: 

If this works for you now, please merge this PR into `refactor/controllers` first so the tests over there pass as well. Then we can merge #39 as well :)

File: src/main/java/edu/kit/provideq/toolbox/meta/SubRoutineDefinition.java [hunk 5]
```diff
@@ -4,9 +4,8 @@
  * Definition of a sub routine that holds basic information.
  *
  * @param type        type of the problem that is used in the sub routine
- * @param url         url of the problem to call the sub routine
  * @param description description of the sub routine call to provide information where and why it is
  *                    needed
  */
-public record SubRoutineDefinition(ProblemType type, String url, String description) {
+public record SubRoutineDefinition(ProblemType type, String description) {
```
Comment (Elscrux) [inline src/main/java/edu/kit/provideq/toolbox/meta/SubRoutineDefinition.java:10]: We still need the url as part of the record. See here for its usage: https://github.com/ProvideQ/toolbox-web/blob/develop/src/components/solvers/SolverPicker.tsx#L123

What we should do here instead is only require the problem type and description in the constructor for `SubRoutineDefinition` so the url part can be inferred from the type.

**Pull request merged**